### PR TITLE
cgroup: support to set threaded mode in cgroup v2

### PR DESCRIPTION
**Pull request by yaoyinnan**
Support to set threaded mode in cgroup v2. The premise of switching to threaded mode is that only the cgroup of cpuset, cpu and pids is supported.

Fixes: #90

Signed-off-by: yaoyinnan <yaoyinnan@foxmail.com>

### CODE REVIEW

File: src/cgroup.rs [hunk 1]
```diff
@@ -328,9 +351,20 @@ fn supported_controllers() -> Vec<String> {
         .collect::<Vec<String>>()
 }
 
-fn create_v2_cgroup(root: PathBuf, path: &str) -> Result<()> {
+fn supported_controllers_from_threaded_mode() -> Vec<String> {
+    let ret = String::from("cpuset cpu pids");
```
Comment (liubin) [inline src/cgroup.rs:355]: Why not juest `vec![String::from("cpuset"), String::from("cpu"), String::from("pids")]` ?

File: src/cgroup.rs [hunk 2]
```diff
@@ -358,6 +457,16 @@ fn create_v2_cgroup(root: PathBuf, path: &str) -> Result<()> {
     Ok(())
 }
 
+pub fn verify_supported_controllers(controllers: &[String]) -> bool {
+    let supported_controllers = get_supported_controllers();
+    for controller in controllers.iter() {
+        if !supported_controllers.contains(controller) {
```
Comment (lifupan) [inline src/cgroup.rs:463]: Is this  check right？

File: src/cgroup.rs [hunk 3]
```diff
@@ -59,19 +59,26 @@ impl Default for Cgroup {
...
                 subsystem.to_controller().create();
             }
         }
     }
 
-    pub fn v2(&self) -> bool {
-        self.hier.v2()
+    /// Create this control group with specified controllers.
+    ///
+    /// Note that this method is only meaningful for cgroup v2.
```
Comment (lifupan) [inline src/cgroup.rs:79]: Why only v2 supported this? May be v1 also need this feature?

File: src/cgroup.rs [hunk 4]
```diff
@@ -241,6 +261,14 @@ impl Cgroup {
         let _ = self.hier.root_control_group().add_task(pid);
     }
 
+    /// Removes a proc from the control group.
+    ///
+    /// Note that this means that the task will be moved back to the root control group in the
+    /// hierarchy and any rules applied to that control group will _still_ apply to the proc.
+    pub fn remove_proc(&self, pid: CgroupPid) {
```
Comment (lifupan) [inline src/cgroup.rs:268]: Why we move the process to root cgroup instead of its parent cgroup?

File: src/cgroup.rs (hunk 3 above)
Comment (gaohuatao-1) [reply src/cgroup.rs:79]: I think that this feature is required in v1 cgroup, too.  In some scenes,  we just want to add the process to those specified subsystems.

**Comment by YushuoEdge** [comment]
Has cgroup v2 threaded mode supported in the upstream cgroup-rs repository?

**Comment by yaoyinnan** [comment]
> Has cgroup v2 threaded mode supported in the upstream cgroup-rs repository?

May I ask which upstream cgroups-rs repository you are referring to?

File: src/cgroup.rs (hunk 3 above)
Comment (YushuoEdge) [reply src/cgroup.rs:79]: Does this mean that only specified controllers support threaded mode, so we should create with specified controller?

File: src/cgroup.rs [hunk 5]
```diff
@@ -265,6 +293,37 @@ impl Cgroup {
             .try_for_each(|sub| sub.to_controller().add_task_by_tgid(&pid))
     }
 
+    /// set cgroup.type
+    pub fn set_cgroup_type(&self, cgroup_type: &str) -> Result<()> {
+        if self.v2() {
+            let subsystems = self.subsystems();
+            if !subsystems.is_empty() {
+                let c = subsystems[0].to_controller();
+                c.set_cgroup_type(cgroup_type)
```
Comment (YushuoEdge) [inline src/cgroup.rs:302]: Only one subsystem needs to be set, and other subsystems are automatically set to cgroup v2？

File: src/cgroup.rs [hunk 6]
```diff
@@ -281,6 +340,33 @@ impl Cgroup {
             .try_for_each(|sub| sub.to_controller().set_release_agent(path))
     }
 
+    /// Returns an Iterator that can be used to iterate over the procs that are currently in the
+    /// control group.
+    pub fn procs(&self) -> Vec<CgroupPid> {
+        // Collect the tasks from all subsystems
```
Comment (YushuoEdge) [inline src/cgroup.rs:346]: typo？tasks or procs?

File: src/cgroup.rs [hunk 7]
```diff
@@ -328,9 +414,22 @@ fn supported_controllers() -> Vec<String> {
...
     // controler list ["memory", "cpu"]
-    let controllers = supported_controllers();
+    let controllers = if let Some(s_controllers) = specified_controllers {
+        if verify_supported_controllers(s_controllers.as_ref()) {
+            s_controllers
+        } else {
+            return Err(Error::new(ErrorKind::SpecifiedControllers));
+        }
+    } else {
+        get_supported_controllers()
```
Comment (YushuoEdge) [inline src/cgroup.rs:430]: If specified controllers contain supported + unsupported controllers, and specfied controllers does not contain all suppored controllers, does this create redundant controllers?

**Review by YushuoEdge — COMMENTED** [review]
Thanks! I have some questions and comments:

**Comment by YushuoEdge** [comment]
> 

levex/cgroup-rs, which this repository is forked from?

**Comment by yaoyinnan** [comment]
> > 
> 
> levex/cgroup-rs, which this repository is forked from?

https://github.com/levex/cgroups-rs
https://github.com/ordovicia/controlgroup-rs of fork from levex has not been updated for a long time, and now the latest cgroups-rs library should be under kata-containers.

**Comment by YushuoEdge** [comment]
> > > 
> > 
> > 
> > levex/cgroup-rs, which this repository is forked from?
> 
> [levex/cgroups-rs](https://github.com/levex/cgroups-rs) and [ordovicia/controlgroup-rs](https://github.com/ordovicia/controlgroup-rs) of fork from levex has not been updated for a long time, and now the latest cgroups-rs library should be under kata-containers.

OK! 👍

File: src/cgroup.rs (hunk 6 above)
Comment (yaoyinnan) [reply src/cgroup.rs:346]: typo

File: src/cgroup.rs (hunk 3 above)
Comment (yaoyinnan) [reply src/cgroup.rs]: Thank you for your review @lifupan @gaohuatao-1 @YushuoEdge. I added the method `load_with_specified_controllers()` for specified section controllers in cgroup v1.

File: src/cgroup.rs (hunk 5 above)
Comment (yaoyinnan) [reply src/cgroup.rs:354]: Yes

File: src/cgroup.rs (hunk 7 above)
Comment (yaoyinnan) [reply src/cgroup.rs:482]: When the specified controller contains an unsupported controller, an error SpecifiedControllers will be returned.
```
#[error("specified controller is not in the list of supported controllers")]
SpecifiedControllers,
```

File: src/cgroup.rs (hunk 4 above)
Comment (yaoyinnan) [reply src/cgroup.rs]: After many tests, it was found that the process could not be moved to the parent control group but could be moved to the root control group. I added the methods move_proc_to_parent and move_task_to_parent to move processes and threads to the parent control group, but kept the method to move to the root control group to ensure that projects using cgroups-rs crate can use remove_proc and remove_task normally.

**Comment by yaoyinnan** [comment]
@lifupan @YushuoEdge @gaohuatao-1 I have processed all the comments, please help me to review again. Thanks.

File: src/cgroup.rs [hunk 8]
```diff
@@ -59,19 +59,30 @@ impl Default for Cgroup {
...
 impl Cgroup {
+    pub fn v2(&self) -> bool {
+        self.hier.v2()
+    }
+
     /// Create this control group.
     fn create(&self) {
         if self.hier.v2() {
-            let _ret = create_v2_cgroup(self.hier.root(), &self.path);
+            let _ret = create_v2_cgroup(self.hier.root(), &self.path, None);
```
Comment (studychao) [inline src/cgroup.rs:69]: `create_v2_cgroup` may be failed but why does the error is dropped?

Comment (yaoyinnan) [reply src/cgroup.rs]: Sorry, fixed.

File: src/cgroup_builder.rs [hunk 9]
```diff
@@ -80,6 +80,7 @@ pub struct CgroupBuilder {
     name: String,
     /// Internal, unsupported field: use the associated builders instead.
     resources: Resources,
+    specified_controllers: Option<Vec<String>>,
```
Comment (studychao) [inline src/cgroup_builder.rs:83]: Cloud you add some comment for this new value in `pub struct` like why do we need a `specified controllers`

**Review by studychao — COMMENTED** [review]
Thanks @yaoyinnan , a few more comments.

File: src/cgroup.rs [hunk 10]
```diff
@@ -233,12 +292,37 @@ impl Cgroup {
...
     /// Note that this means that the task will be moved back to the root control group in the
     /// hierarchy and any rules applied to that control group will _still_ apply to the task.
-    pub fn remove_task(&self, pid: CgroupPid) {
-        let _ = self.hier.root_control_group().add_task(pid);
+    pub fn remove_task(&self, pid: CgroupPid) -> Result<()> {
+        self.hier.root_control_group().add_task(pid)
+    }
+
+    /// Move to a proc to parent control group.
+    pub fn move_proc_to_parent(&self, pid: CgroupPid) -> Result<()> {
```
[the remark this answers is not in the document]
Comment (yaoyinnan) [reply src/cgroup.rs]: Yes, it is more accurate to use tgid and tid here.

File: src/cgroup.rs [hunk 11]
```diff
@@ -260,9 +344,50 @@ impl Cgroup {
 
     /// Attach a task to the control group by thread group id.
     pub fn add_task_by_tgid(&self, pid: CgroupPid) -> Result<()> {
```
[the remark this answers is not in the document]
Comment (yaoyinnan) [reply src/cgroup.rs]: Yes, it is more accurate to use tgid and tid here.

File: src/cgroup_builder.rs (hunk 9 above)
Comment (yaoyinnan) [reply src/cgroup_builder.rs:84]: added

**Review by studychao — APPROVED** [review]
LGTM

File: src/cgroup.rs [hunk 12]
```diff
@@ -59,43 +59,76 @@ impl Default for Cgroup {
...
+            Ok(())
         }
     }
 
-    pub fn v2(&self) -> bool {
-        self.hier.v2()
+    /// Create this control group with specified controllers.
+    fn create_with_specified_controllers(
+        &self,
+        specified_controllers: Option<Vec<String>>,
```
Comment (YushuoEdge) [inline src/cgroup.rs:81]: The variable `specified_controllers` seems only work with Cgroup V2?
I wonder:

- Will this function be called when using Cgroup v1?
- If yes, what will happen if `specified_controllers` do not match with `self.subsystems`?
- If no, Why we do not make judgments in new and load?

Comment (yaoyinnan) [reply src/cgroup.rs]: Yes, cgroup v1 also requires filtering on specified controllers. I modified to filter in new and load.

**Review by YushuoEdge — APPROVED** [review]
LGTM!

**Review by liubin — APPROVED** [review]

**Comment by Tim-Zhang** [comment]
@yaoyinnan This patch has been released in version [0.2.12](https://crates.io/crates/cgroups-rs/0.2.12), thanks

**Pull request merged**